OpenRA.Mods.Common/Effects/FloatingSprite.cs

```diff
@@ -0,0 +1,98 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Effects;
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.Effects
+{
+	class FloatingSprite : IEffect, ISync, ISpatiallyPartitionable
+	{
+		readonly WDist[] speed;
+		readonly WDist[] gravity;
+		readonly Animation anim;
+
+		readonly bool visibleThroughFog;
+		readonly int turnRate;
+		readonly int randomRate;
+		readonly string palette;
+
+		[Sync]
+		WPos pos;
+		WVec offset;
+		int lifetime;
+		int ticks;
+		int facing;
+
+		public FloatingSprite(Actor emitter, string image, string[] sequences, string palette, bool isPlayerPalette,
+			int[] duration, WDist[] speed, WDist[] gravity, int turnRate, int randomRate, WPos pos, int facing = -1,
+			bool visibleThroughFog = false)
+		{
+			var world = emitter.World;
+			this.pos = pos;
+			this.turnRate = turnRate;
+			this.randomRate = randomRate;
+			this.speed = speed;
+			this.gravity = gravity;
+			this.visibleThroughFog = visibleThroughFog;
+
+			this.facing = facing > -1
+				? facing
+				: world.SharedRandom.Next(256);
+
+			anim = new Animation(world, image, () => WAngle.FromFacing(facing));
```

Comment:
Maybe it is better to pass the facing as `WAngle` directly instead of converting wangle -> int -> wangle? @pchote can probably comment on what is best to use here.

Reply:
Changed it, but I still do some `int` to `WAngle` conversions so there seems no clear benefit.